mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala

```diff
@@ -290,6 +293,27 @@ object Bucketizer extends DefaultParamsReadable[Bucketizer] {
     }
   }
 
+
+  private[Bucketizer] class BucketizerWriter(instance: Bucketizer) extends MLWriter {
+
+    override protected def saveImpl(path: String): Unit = {
+      // SPARK-23377: The default params will be saved and loaded as user-supplied params.
+      // Once `inputCols` is set, the default value of `outputCol` param causes the error
+      // when checking exclusive params. As a temporary to fix it, we remove the default
+      // value of `outputCol` if `inputCols` is set before saving.
+      // TODO: If we modify the persistence mechanism later to better handle default params,
+      // we can get rid of this.
+      var removedOutputCol: Option[String] = None
```

Comment:
I doubt whether it need a "lock" here, because it is the way "clear default value first, then save model, then restore default value".
Maybe wrapping the code block here by `synchronized` is safer ?

Reply:
I was thinking about this too. But looks like we don't add lock to the places we change params in ML. I guess that we assume the usage of ML models is single-threaded. So I leave it as this. Will add it if others think this is required too.

Reply:
yep. But I have some new thoughts, see my comments at bottom. -:)